### PR TITLE
perf: cache default session contexts

### LINKED ISSUE #2705: perf: re-use session context
We create a new session context every time we call `execute_plan` (used in queries and merge insert):

https://github.com/lancedb/lance/blob/293758e41acb57dfa61807bf96802edf39ee0cc9/rust/lance-datafusion/src/exec.rs#L214-L218

This ends up becoming a meaningful amount of time in query execution:

image

We should instead create a single session and re-use it across queries.

### PR DESCRIPTION
**Pull request by dsgibbons**
Closes #2705.

This PR uses `lazy_static` to introduce two default session contexts:

- `DEFAULT_SESSION_CONTEXT`
- `DEFAULT_SESSION_CONTEXT_WITH_SPILLING`

Users request the session context via the new `get_session_context` function. If the user requests a session context with the default memory pool size, then one of the above contexts are returned. Otherwise, we compute a new session context like before.

### CODE REVIEW

**Review by wjones127 — COMMENTED** [review]
We have a notion of a `Session` here, which holds various in-memory caches right now. That seems like a good place for this:

https://github.com/lancedb/lance/blob/a7e65f4a7570910aa9ce24ac17824ae3c16c1918/rust/lance/src/session.rs#L21-L30

Might have to move around the `DeepSizeOf` implementation to just be for some cache property, but I think this would be a good place to hold the session context.

**Comment by westonpace** [comment]
Thanks for working on this!

> There may be good reason not to embed state inside LanceExecutionOptions - if this is not the right place for it, where should the session be cached?

From a semantics standpoint I, as a user, would not expect something named "options" to be any more than a bag of config settings and I wouldn't expect it to have state.

I'm also not sure this fully solves the problem because I think, right now, we create a new `LanceExecutionOptions` every time we need one.

My thinking is that we will eventually want some kind of process-wide singleton that we use if the caller hasn't requested something special.  We can use `lazy_static` to get this (once Rust 1.80 is our MSRV we can use `LazyCell` from `std` too).

> We have a notion of a Session here, which holds various in-memory caches right now. That seems like a good place for this:

Is there any particular reason we would want each Dataset to have its own `SessionContext`?  It looks like the session context holds:

* UDFs - we don't use and, if we did, would be consistent across all datasets
* Catalog (e.g. registered data soruces) - we don't use, hard to know if it would be consistent or not, not sure why we would use it
* Optimizer rules / analyzer rules - we do use these, but they are consistent across all datasets

**Comment by dsgibbons** [comment]
> From a semantics standpoint I, as a user, would not expect something named "options" to be any more than a bag of config settings and I wouldn't expect it to have state.

Makes sense, I'm happy to put it somewhere else.

> We have a notion of a Session here, which holds various in-memory caches right now. 

It's not obvious to me how `Session` would propagate to `execute_plan` without changing the `execute_plan` API. We would either need an additional trait for the first argument (something like `plan: Arc<dyn ExecutionPlan + WithSession>`) or an additional argument (e.g., `session: Option<SessionContext>`). I've seen something like the latter done in a few other places too (e.g., https://github.com/lancedb/lance/blob/main/rust/lance-file/src/reader.rs#L92). Are we OK with changing the `execute_plan` API so we can pull out the cached `SessionContext`?

> My thinking is that we will eventually want some kind of process-wide singleton that we use if the caller hasn't requested something special. We can use lazy_static to get this (once Rust 1.80 is our MSRV we can use LazyCell from std too).

This was my first thought, but I wasn't sure how the team would feel about introducing a global variable. Are you OK with this option @wjones127?

**Comment by wjones127** [comment]
> I wasn't sure how the team would feel about introducing a global variable. Are you OK with this option?

Yes, I think that's a sensible solution. 👍

**Comment by dsgibbons** [comment]
Updated with default `SessionContext`s for `use_spilling=false` and `use_spilling=true`.

**Comment by dsgibbons** [comment]
Current solution does not properly account for `LANCE_BYPASS_SPILLING` and `LANCE_MEM_POOL_SIZE` environment variables - hence why the Python tests are failing (I had only tested locally with Rust). I'll get this fixed.

**Comment by dsgibbons** [comment]
Environment variables are now accounted for. 

> some kind of process-wide singleton that we use if the caller hasn't requested something special.

I'm using the cached `SessionContext` if `mem_pool_size=DEFAULT_LANCE_MEM_POOL_SIZE`. We may eventually want a map that is updated with more specific user requests (e.g., different `mem_pool_size`). However, this would require adding something like `RwLock` to `lazy_static` to enable interior mutability.

**Review by wjones127 — CHANGES_REQUESTED** [review]
There is one place in `merge_insert.rs` where we use `new_session_context`, but I think we should instead re-use `DEFAULT_SESSION_CONTEXT_WITH_SPILLING`. Could we do that in this PR as well? Otherwise this looks good to me.

**Comment by dsgibbons** [comment]
> There is one place in merge_insert.rs where we use new_session_context, but I think we should instead re-use DEFAULT_SESSION_CONTEXT_WITH_SPILLING. Could we do that in this PR as well? Otherwise this looks good to me.

Done via adding a `get_session_context` public function.

File: rust/lance-datafusion/src/exec.rs [hunk 1]
```diff
@@ -208,14 +209,40 @@ pub fn new_session_context(options: LanceExecutionOptions) -> SessionContext {
...
+        new_session_context(LanceExecutionOptions {
+            use_spilling: true,
+            ..Default::default()
+        })
+    };
+}
+
+pub fn get_session_context(options: LanceExecutionOptions) -> SessionContext {
+    let session_ctx: SessionContext;
+    if options.mem_pool_size() == DEFAULT_LANCE_MEM_POOL_SIZE {
```
Comment (wjones127) [inline rust/lance-datafusion/src/exec.rs:225]: IIUC, if you set a custom memory pool size different from the default, then you will end up creating a new session context each time this is called, won't you?

Comment (dsgibbons) [reply rust/lance-datafusion/src/exec.rs:225]: That is correct - that design decision was intentional based on the comment from @westonpace: 

> My thinking is that we will eventually want some kind of process-wide singleton that we use if the caller hasn't requested something special.

I interpreted "something special" as a custom memory pool size. We can make this more generic by introducing a `HashMap`, but we would then need to make the `lazy_static` ref mutable.

Comment (wjones127) [reply rust/lance-datafusion/src/exec.rs:225]: Got it. Let's create a follow-up issue, but this already seems like an improvement.

**Review by wjones127 — APPROVED** [review]

File: rust/lance-datafusion/src/exec.rs (hunk 1 above)
Comment (dsgibbons) [reply rust/lance-datafusion/src/exec.rs:225]: PR description updated so the implementation is more obvious.

**Pull request merged**